### PR TITLE
cmake: Add simple recursive ctags target for Ceph source only

### PR DESCRIPTION
**Pull request by tchaikov**
Excludes submodules and random JavaScript.

Signed-off-by: Dan Mick <dan.mick@redhat.com>
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
i don't use ctags. so it **should** work in theory.

**Comment by dmick** [comment]
This resulted in (from ps):

/usr/bin/ctags -R --c++-kinds=+p --fields=+iaS --extra=+q --exclude=*.js --exclude={src/civetweb,src/erasure-code/jerasure/jerasure,src/erasure-code/jerasure/gf-complete,src/rocksdb,src/googletest,src/spdk,src/xxHash,src/isa-l,src/lua,src/Beast,src/boost,src/dpdk,src/zstd} src

1) it's not clear to me that --exclude can take a {} pattern, but it wasn't expanded by the shell, anyway
2) the extra src argument seems wrong, or at least I don't understand it
3) this is building in srcdir, not srcdir/src, which I think is undesirable

**Comment by dmick** [comment]
something like
    COMMAND awk "/${src_dir}/ { print gsub(\"src/\", \"\", $2 }"

but with the quoting mess figured out, perhaps

**Comment by dmick** [comment]
it's a *bash* wildcard pattern, but I'm happy it works.

File: cmake/modules/CTags.cmake [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  set(tag_file ${src_dir}/tags)
+  execute_process(
+    COMMAND git config --file .gitmodules --get-regexp path
+    COMMAND awk "/${src_dir}/ { print $2 }"
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    OUTPUT_VARIABLE submodules
+    OUTPUT_STRIP_TRAILING_WHITESPACE)
+  string(REPLACE "\n" " " submodules ${submodules})
+  string(REPLACE "${src_dir}/" "--exclude=" excludes ${submodules})
+  separate_arguments(excludes UNIX_COMMAND ${excludes})
```
Comment (dmick) [inline cmake/modules/CTags.cmake]: I agree this is probably minimal, but it's extremely confusing even if you're aware of cmake's Lists and automagic shell-metacharacter quoting that includes spaces...so let's add a comment like this above it:

```
  # cmake will quote any embedded spaces when expanding ${excludes}
  # in the command below, so, turn ${excludes} into a cmake List (separated
  # by semicolons) to let the expansion below simply replace them with spaces
  # again.
```

File: cmake/modules/CTags.cmake [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  add_custom_target(ctags
+    COMMAND ${CTAGS_EXECUTABLE} -R --c++-kinds=+p --fields=+iaS --extra=+q --exclude=*.js ${excludes}
+    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/${src_dir}
+    COMMENT "Building ctags file ${tag_file} (no submodules)"
+    VERBATIM)
+  set_source_files_properties(${CMAKE_SOURCE_DIR}/${tag_file} PROPERTIES
```
Comment (dmick) [inline cmake/modules/CTags.cmake]: ...and let's also add the 'tags' target depending on 'ctags'

**Comment by tchaikov** [comment]
@dmick comments addressed and fix repushed.

**Comment by cbodley** [comment]
just a general comment, but i do find it useful to have tags for headers in submodules (my current `~/.ctags` only excludes `boost`). i guess i can use this feature to `make ctags`, then `--append` some submodules manually

**Comment by dmick** [comment]
@cbodley I certainly have no compunction with parameterizing this, perhaps based on an env var, or perhaps a new target ctags-sub or something

**Comment by dmick** [comment]
Test looks good today.  Thanks @tchaikov

**Pull request merged**